core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -2626,8 +2601,16 @@ class ReplicaManager(val config: KafkaConfig,
     localLeaders.forKeyValue { (tp, info) =>
       getOrCreatePartition(tp, delta, info.topicId).foreach { case (partition, isNew) =>
         try {
+          val partitionDirectoryId: Option[Uuid] =
+            Option(newImage.topics().getPartition(info.topicId(), partition.partitionId)).map(_.directory(localBrokerId))
+
           val state = info.partition.toLeaderAndIsrPartitionState(tp, isNew)
-          partition.makeLeader(state, offsetCheckpoints, Some(info.topicId))
+          partition.makeLeader(state, offsetCheckpoints, Some(info.topicId), partitionDirectoryId)
+
+          if (!delta.localChanges(localBrokerId).directoryIds().isEmpty) {
+            maybeUpdateTopicAssignment(partition, partitionDirectoryId)
+          }
```

Comment:
This can be pulled up to `def applyDelta(delta: TopicsDelta, newImage: MetadataImage):`.
Maybe we can do `localChanges.directoryIds().forEach(maybeUpdateTopicAssignment)` after 
```
        if (!localChanges.leaders.isEmpty) {
          applyLocalLeadersDelta(leaderChangedPartitions, newImage, delta, lazyOffsetCheckpoints, localChanges.leaders.asScala)
        }
        if (!localChanges.followers.isEmpty) {
          applyLocalFollowersDelta(followerChangedPartitions, newImage, delta, lazyOffsetCheckpoints, localChanges.followers.asScala)
        }
        localChanges.directoryIds().forEach(maybeUpdateTopicAssignment)
```

WDYT?

Reply:
We need to update the assignment with the actual directory ID of the partition created at the end not with `directoryIds`. This info is available in `Partition` (not `TopicPartition`) which we don't have at this point.

Reply:
`applyLocalLeadersDelta` and `applyLocalFollowersDelta` are only called if `localChanges.leaders`/`localChanges.partitions` are not empty.
I think `localChanges.directoryIds` should be considered separately.

Reply:
I updated this now. also updated `directoryIds` to return `TopicIdParititon` instead of `TopicPartition`